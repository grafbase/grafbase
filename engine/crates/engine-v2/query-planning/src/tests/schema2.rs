use crate::{
    tests::{read_schema, TestOperation},
    OperationGraph, Solver,
};

const SCHEMA: &str = r###"
directive @core(feature: String!) repeatable on SCHEMA

directive @join__owner(graph: join__Graph!) on OBJECT

directive @join__type(
    graph: join__Graph!
    key: String!
    resolvable: Boolean = true
) repeatable on OBJECT | INTERFACE

directive @join__field(
    graph: join__Graph
    requires: String
    provides: String
) on FIELD_DEFINITION

directive @join__graph(name: String!, url: String!) on ENUM_VALUE

directive @join__implements(graph: join__Graph!, interface: String!) repeatable on OBJECT | INTERFACE

directive @join__unionMember(graph: join__Graph!, member: String!) repeatable on UNION

enum join__Graph {
    ACCOUNTS @join__graph(name: "accounts", url: "http://127.0.0.1:39203/")
    INVENTORY @join__graph(name: "inventory", url: "http://127.0.0.1:33259/")
    PRODUCTS @join__graph(name: "products", url: "http://127.0.0.1:42663/")
    REVIEWS @join__graph(name: "reviews", url: "http://127.0.0.1:38447/")
    SHIPPING @join__graph(name: "shipping", url: "http://127.0.0.1:38213/")
}

type BusinessAccount implements Account
    @join__implements(graph: SHIPPING, interface: "Account")
    @join__type(graph: ACCOUNTS, key: "id")
    @join__type(graph: SHIPPING, key: "id email joinedTimestamp", resolvable: false)
{
    businessName: String! @join__field(graph: ACCOUNTS)
    email: String! @join__field(graph: ACCOUNTS)
    id: ID!
    joinedTimestamp: Int! @join__field(graph: ACCOUNTS)
}

type Cart {
    products: [Product!]! @join__field(graph: ACCOUNTS)
}

type Picture {
    altText: String! @inaccessible
    height: Int!
    url: String!
    width: Int!
}

type Product
    @join__type(graph: ACCOUNTS, key: "name", resolvable: false)
    @join__type(graph: INVENTORY, key: "upc")
    @join__type(graph: PRODUCTS, key: "upc")
    @join__type(graph: PRODUCTS, key: "name")
    @join__type(graph: REVIEWS, key: "upc")
{
    availableShippingService: [ShippingService!]! @join__field(graph: INVENTORY, requires: "weight(unit: KILOGRAM)")
    name: String!
    price: Int! @join__field(graph: PRODUCTS)
    reviews: [Review!]! @join__field(graph: REVIEWS)
    shippingEstimate: Int! @join__field(graph: INVENTORY, requires: "weight(unit: KILOGRAM)")
    upc: String!
    weight(unit: WeightUnit!): Float! @join__field(graph: PRODUCTS)
}

type User implements Account
    @join__implements(graph: SHIPPING, interface: "Account")
    @join__type(graph: ACCOUNTS, key: "id")
    @join__type(graph: REVIEWS, key: "id")
    @join__type(graph: SHIPPING, key: "id username reviewCount joinedTimestamp", resolvable: false)
{
    cart: Cart! @join__field(graph: ACCOUNTS)
    id: ID!
    joinedTimestamp: Int! @join__field(graph: ACCOUNTS)
    profilePicture: Picture @join__field(graph: ACCOUNTS)
    """
    This used to be part of this subgraph, but is now being overridden from
    `reviews`
    """
    reviewCount: Int! @join__field(graph: REVIEWS, override: "accounts")
    reviews: [Review!]! @join__field(graph: REVIEWS)
    trustworthiness: Trustworthiness! @join__field(graph: REVIEWS, requires: "joinedTimestamp")
    username: String! @join__field(graph: ACCOUNTS)
}

type DeliveryCompany implements ShippingService & ShippingModality
    @join__implements(graph: INVENTORY, interface: "ShippingService")
    @join__implements(graph: SHIPPING, interface: "ShippingModality")
    @join__type(graph: INVENTORY, key: "id")
    @join__type(graph: SHIPPING, key: "id")
{
    companyType: String! @join__field(graph: INVENTORY)
    id: String!
    name: String! @join__field(graph: INVENTORY)
    qualifiedName: String! @join__field(graph: SHIPPING, requires: "companyType")
    reviews: [ShippingServiceReview!]! @join__field(graph: REVIEWS, requires: "name")
}

type HomingPigeon implements ShippingService & ShippingModality
    @join__implements(graph: INVENTORY, interface: "ShippingService")
    @join__implements(graph: SHIPPING, interface: "ShippingModality")
    @join__type(graph: INVENTORY, key: "id")
    @join__type(graph: SHIPPING, key: "id")
{
    id: String!
    name: String! @join__field(graph: INVENTORY)
    nickname: String! @join__field(graph: INVENTORY)
    qualifiedName: String! @join__field(graph: SHIPPING, requires: "nickname")
    reviews: [ShippingServiceReview!]! @join__field(graph: REVIEWS, requires: "name")
}

type Review {
    author: User @join__field(graph: REVIEWS)
    body: String! @join__field(graph: REVIEWS)
    id: ID! @join__field(graph: REVIEWS)
    pictures: [Picture!]! @join__field(graph: REVIEWS)
    product: Product! @join__field(graph: REVIEWS, provides: "price")
}

type ShippingServiceReview {
    body: String! @join__field(graph: REVIEWS)
}

type ShippingOptions {
    defaultCompanySummary: String! @join__field(graph: SHIPPING, requires: "defaultDeliveryCompany { companyType }")
    defaultDeliveryCompany: DeliveryCompany! @join__field(graph: SHIPPING)
    modalities: [ShippingModality!]! @join__field(graph: SHIPPING)
    seller: Account! @join__field(graph: SHIPPING, provides: "... on BusinessAccount { email }... on User { reviewCount }")
    summary: String! @join__field(graph: SHIPPING, requires: "modalities { ... on HomingPigeon { nickname }... on DeliveryCompany { companyType } }")
}

type Query {
    me: User! @join__field(graph: ACCOUNTS)
    product(upc: String!): Product @join__field(graph: PRODUCTS)
    shippingOptions: ShippingOptions! @join__field(graph: SHIPPING)
    topProducts: [Product!]! @join__field(graph: PRODUCTS)
}

type Subscription {
    newProducts: Product! @join__field(graph: PRODUCTS)
}

interface ShippingService
    @join__type(graph: INVENTORY, key: "id")
    @join__type(graph: REVIEWS, key: "id", isInterfaceObject: true)
{
    id: String!
    name: String! @join__field(graph: INVENTORY)
    reviews: [ShippingServiceReview!]! @join__field(graph: REVIEWS, requires: "name")
}

interface Account {
    id: ID!
    joinedTimestamp: Int!
}

interface ShippingModality {
    id: String!
    name: String!
    qualifiedName: String!
}

enum WeightUnit {
    KILOGRAM
    GRAM
}

enum Trustworthiness {
    REALLY_TRUSTED
    KINDA_TRUSTED
    NOT_TRUSTED
}
"###;

#[test]
fn sibling_dependency() {
    let schema = read_schema(SCHEMA);
    let mut operation = TestOperation::bind(
        &schema,
        r#"
        query {
            me {
                id
                username
                cart {
                    products {
                        price
                        reviews {
                            author {
                                id
                                username
                            }
                            body
                        }
                    }
                }
            }
        }
    "#,
    );

    let mut graph = OperationGraph::new(&schema, &mut operation).unwrap();
    insta::assert_snapshot!("graph", graph.to_dot_graph(), &graph.to_pretty_dot_graph());

    let mut solver = Solver::initialize(&graph).unwrap();
    insta::assert_snapshot!("solver", solver.to_dot_graph(), &solver.to_pretty_dot_graph());

    solver.execute().unwrap();
    insta::assert_snapshot!("solved", solver.to_dot_graph(), &solver.to_pretty_dot_graph());

    graph.finalize_operation_graph_with_solution(solver.into_solution());
    insta::assert_snapshot!("solved-graph", graph.to_dot_graph(), &graph.to_pretty_dot_graph());
}
