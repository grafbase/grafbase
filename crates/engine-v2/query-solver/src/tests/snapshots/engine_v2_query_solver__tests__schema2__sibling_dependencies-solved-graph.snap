---
source: engine/crates/engine-v2/query-planning/src/tests/schema2.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"id\" ]\n    2 [ label = \"username\" ]\n    3 [ label = \"price\" ]\n    4 [ label = \"id\" ]\n    5 [ label = \"username\" ]\n    6 [ label = \"author\" ]\n    7 [ label = \"body\" ]\n    8 [ label = \"reviews\" ]\n    9 [ label = \"products\" ]\n    10 [ label = \"cart\" ]\n    11 [ label = \"me\" ]\n    12 [ label = \"Root#accounts\", shape=parallelogram, color=dodgerblue ]\n    13 [ label = \"me@Root#accounts\", shape=box, color=dodgerblue ]\n    14 [ label = \"cart@Root#accounts\", shape=box, color=dodgerblue ]\n    15 [ label = \"products@Root#accounts\", shape=box, color=dodgerblue ]\n    16 [ label = \"FedEntity#reviews\", shape=parallelogram, color=dodgerblue ]\n    17 [ label = \"reviews@FedEntity#reviews\", shape=box, color=dodgerblue ]\n    18 [ label = \"body@FedEntity#reviews\", shape=box, color=dodgerblue ]\n    19 [ label = \"author@FedEntity#reviews\", shape=box, color=dodgerblue ]\n    20 [ label = \"FedEntity#accounts\", shape=parallelogram, color=dodgerblue ]\n    21 [ label = \"username@FedEntity#accounts\", shape=box, color=dodgerblue ]\n    22 [ label = \"id@FedEntity#reviews\", shape=box, color=dodgerblue ]\n    25 [ label = \"FedEntity#products\", shape=parallelogram, color=dodgerblue ]\n    26 [ label = \"price@FedEntity#products\", shape=box, color=dodgerblue ]\n    27 [ label = \"username@Root#accounts\", shape=box, color=dodgerblue ]\n    28 [ label = \"id@Root#accounts\", shape=box, color=dodgerblue ]\n    29 [ label = \"*name\" ]\n    30 [ label = \"name@Root#accounts\", shape=box, color=dodgerblue ]\n    32 [ label = \"*upc\" ]\n    33 [ label = \"upc@FedEntity#products\", shape=box, color=dodgerblue ]\n    0 -> 11 [ label = \"\" ]\n    0 -> 12 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    0 -> 12 [ label = \"\", style=dashed,arrowhead=none ]\n    12 -> 13 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    13 -> 11 [ label = \"\", color=violet,arrowhead=none ]\n    11 -> 10 [ label = \"\" ]\n    13 -> 14 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    14 -> 10 [ label = \"\", color=violet,arrowhead=none ]\n    10 -> 9 [ label = \"\" ]\n    14 -> 15 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    15 -> 9 [ label = \"\", color=violet,arrowhead=none ]\n    9 -> 8 [ label = \"\" ]\n    15 -> 16 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    9 -> 16 [ label = \"\", style=dashed,arrowhead=none ]\n    16 -> 17 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    17 -> 8 [ label = \"\", color=violet,arrowhead=none ]\n    8 -> 7 [ label = \"\" ]\n    17 -> 18 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    18 -> 7 [ label = \"\", color=violet,arrowhead=none ]\n    8 -> 6 [ label = \"\" ]\n    17 -> 19 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    19 -> 6 [ label = \"\", color=violet,arrowhead=none ]\n    6 -> 5 [ label = \"\" ]\n    19 -> 20 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    6 -> 20 [ label = \"\", style=dashed,arrowhead=none ]\n    20 -> 21 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    21 -> 5 [ label = \"\", color=violet,arrowhead=none ]\n    6 -> 4 [ label = \"\" ]\n    19 -> 22 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    22 -> 4 [ label = \"\", color=violet,arrowhead=none ]\n    9 -> 3 [ label = \"\" ]\n    15 -> 25 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    9 -> 25 [ label = \"\", style=dashed,arrowhead=none ]\n    25 -> 26 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    26 -> 3 [ label = \"\", color=violet,arrowhead=none ]\n    11 -> 2 [ label = \"\" ]\n    13 -> 27 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    27 -> 2 [ label = \"\", color=violet,arrowhead=none ]\n    11 -> 1 [ label = \"\" ]\n    13 -> 28 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    28 -> 1 [ label = \"\", color=violet,arrowhead=none ]\n    25 -> 29 [ label = \"\", color=orangered,arrowhead=inv ]\n    9 -> 29 [ label = \"\" ]\n    15 -> 30 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    30 -> 29 [ label = \"\", color=violet,arrowhead=none ]\n    9 -> 32 [ label = \"\" ]\n    25 -> 33 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    33 -> 32 [ label = \"\", color=violet,arrowhead=none ]\n    20 -> 4 [ label = \"\", color=orangered,arrowhead=inv ]\n    16 -> 32 [ label = \"\", color=orangered,arrowhead=inv ]\n}\n"
---
digraph {
    0 [ root]
    1 [ id]
    2 [ username]
    3 [ price]
    4 [ id]
    5 [ username]
    6 [ author]
    7 [ body]
    8 [ reviews]
    9 [ products]
    10 [ cart]
    11 [ me]
    12 [ Root#accounts]
    13 [ me@Root#accounts]
    14 [ cart@Root#accounts]
    15 [ products@Root#accounts]
    16 [ FedEntity#reviews]
    17 [ reviews@FedEntity#reviews]
    18 [ body@FedEntity#reviews]
    19 [ author@FedEntity#reviews]
    20 [ FedEntity#accounts]
    21 [ username@FedEntity#accounts]
    22 [ id@FedEntity#reviews]
    25 [ FedEntity#products]
    26 [ price@FedEntity#products]
    27 [ username@Root#accounts]
    28 [ id@Root#accounts]
    29 [ *name]
    30 [ name@Root#accounts]
    32 [ *upc]
    33 [ upc@FedEntity#products]
    0 -> 11 [ label = "Field" ]
    0 -> 12 [ label = "CreateChildResolver" ]
    0 -> 12 [ label = "HasChildResolver" ]
    12 -> 13 [ label = "CanProvide" ]
    13 -> 11 [ label = "Provides" ]
    11 -> 10 [ label = "Field" ]
    13 -> 14 [ label = "CanProvide" ]
    14 -> 10 [ label = "Provides" ]
    10 -> 9 [ label = "Field" ]
    14 -> 15 [ label = "CanProvide" ]
    15 -> 9 [ label = "Provides" ]
    9 -> 8 [ label = "Field" ]
    15 -> 16 [ label = "CreateChildResolver" ]
    9 -> 16 [ label = "HasChildResolver" ]
    16 -> 17 [ label = "CanProvide" ]
    17 -> 8 [ label = "Provides" ]
    8 -> 7 [ label = "Field" ]
    17 -> 18 [ label = "CanProvide" ]
    18 -> 7 [ label = "Provides" ]
    8 -> 6 [ label = "Field" ]
    17 -> 19 [ label = "CanProvide" ]
    19 -> 6 [ label = "Provides" ]
    6 -> 5 [ label = "Field" ]
    19 -> 20 [ label = "CreateChildResolver" ]
    6 -> 20 [ label = "HasChildResolver" ]
    20 -> 21 [ label = "CanProvide" ]
    21 -> 5 [ label = "Provides" ]
    6 -> 4 [ label = "Field" ]
    19 -> 22 [ label = "CanProvide" ]
    22 -> 4 [ label = "Provides" ]
    9 -> 3 [ label = "Field" ]
    15 -> 25 [ label = "CreateChildResolver" ]
    9 -> 25 [ label = "HasChildResolver" ]
    25 -> 26 [ label = "CanProvide" ]
    26 -> 3 [ label = "Provides" ]
    11 -> 2 [ label = "Field" ]
    13 -> 27 [ label = "CanProvide" ]
    27 -> 2 [ label = "Provides" ]
    11 -> 1 [ label = "Field" ]
    13 -> 28 [ label = "CanProvide" ]
    28 -> 1 [ label = "Provides" ]
    25 -> 29 [ label = "Requires" ]
    9 -> 29 [ label = "Field" ]
    15 -> 30 [ label = "CanProvide" ]
    30 -> 29 [ label = "Provides" ]
    9 -> 32 [ label = "Field" ]
    25 -> 33 [ label = "CanProvide" ]
    33 -> 32 [ label = "Provides" ]
    20 -> 4 [ label = "Requires" ]
    16 -> 32 [ label = "Requires" ]
}
